### PR TITLE
test(core): signals tests

### PR DESCRIPTION
**Pull request by tltsutltsu**
Resolves: #2846

@gear-tech/dev

### CODE REVIEW

File: examples/signal-entry/src/lib.rs [hunk 1]
```diff
@@ -221,6 +253,17 @@ mod wasm {
                 msg::send_bytes((*user).into(), b"handle_signal_forbidden_call", 0).unwrap();
                 let _ = msg::source();
             }
+            HandleSignalState::Assert(signal_saved) => {
+                let signal_received = msg::signal_code()
+                    .expect("Incorrect call")
+                    .expect("Unsupported code");
+
+                if signal_received == *signal_saved {
+                    msg::send(unsafe { INITIATOR }, b"handle_signal_equal", 0).unwrap();
```
Comment (breathx) [inline examples/signal-entry/src/lib.rs:262]: isn't true/false sending easier?

File: pallets/gear/src/tests.rs [hunk 2]
```diff
@@ -14095,4 +14096,56 @@ mod utils {
...
+                WASM_BINARY.to_vec(),
+                DEFAULT_SALT.to_vec(),
+                USER_1.encode(),
+                10_000_000_000,
+                0,
+            ));
+
+            let pid = get_last_program_id();
+
+            run_to_block(2, None);
```
Comment (breathx) [inline pallets/gear/src/tests.rs:14121]: assert program is active and initialised

File: pallets/gear/src/tests.rs [hunk 3]
```diff
@@ -14095,4 +14096,56 @@ mod utils {
...
+                action.encode(),
+                10_000_000_000,
+                0,
+            ));
+
+            let mid = get_last_message_id();
+
+            run_to_block(4, None);
+
+            assert!(GasHandlerOf::<Test>::get_system_reserve(mid).is_err());
```
Comment (breathx) [inline pallets/gear/src/tests.rs:14145]: assert it exists before

**Review by breathx — CHANGES_REQUESTED** [review]
Please write somewhere docs for each case and ref tests that we covered all of them
Start with general cases and extend with their subcases (e.g. panic case: oom, out of gas, anything else ... )

**Comment by StackOverflowExcept1on** [comment]
Something wrong happen with branch. Github shows +-5000 lines changed in this PR.

**Comment by breathx** [comment]
Awesome doc

**Review by clearloop — APPROVED** [review]

File: core-errors/signal-code-testing.md [hunk 4]
```diff
@@ -0,0 +1,1059 @@
...
+
+        // Ensure that signal code sent is signal code we saved
+        let mail_msg = get_last_mail(USER_1);
+        assert_eq!(mail_msg.payload_bytes(), true.encode());
+    }
+    ```
+    </details>
+- Gas runs out during plain or lazy pages memory access.
+
+    This case is quite tricky to test, because accurately predicting the amount of gas consumed during program execution and estimating gas usage upon memory access can be difficult. As a result, the test for this case is not provided here.
```
[the remark this answers is not in the document]
Comment (tltsutltsu) [reply core-errors/signal-code-testing.md]: that's a good way, thanks!

**Pull request merged**